amazon/ion/writer_binary_raw_fields.py

```diff
@@ -0,0 +1,240 @@
+        return 1
+    return uncached_func(buf, value)
+
+
+def _write_base(buf, value, bits_per_octet, end_bit=0, sign_bit=0, is_signed=False):
+    """Write a field to the provided buffer.
+
+    Args:
+    buf (Sequence): The buffer into which the UInt will be written, in the form of integer octets.
+    value (int): The value to write as a UInt.
+    bits_per_octet (int): The number of value bits (i.e. exclusive of the end bit, but inclusive of the sign bit,
+                          if applicable) per octet.
+    end_bit (Optional[int]): The end bit mask.
+    sign_bit (Optional[int]): The sign bit mask.
+
+    Returns:
+        int: The number of octets written.
+    """
+    if value == 0:
+        buf.append(sign_bit | end_bit)
+        return 1
+    num_bits = value.bit_length()
+    num_octets = num_bits // bits_per_octet
+    # 'remainder' is the number of value bits in the first octet.
+    remainder = num_bits % bits_per_octet
+    if remainder != 0 or is_signed:  # If signed, the first octet has one fewer bit available, requiring another octet.
+        num_octets += 1
+    else:
+        remainder = bits_per_octet  # This ensures that unsigned values that fit exactly are not shifted too far.
+    for i in range(num_octets):
+        octet = 0
+        if i == 0:
+            octet |= sign_bit
+        if i == num_octets - 1:
+            octet |= end_bit
+        # 'remainder' is used for alignment such that only the first octet may contain insignificant zeros.
+        octet |= ((value >> (num_bits - (remainder + bits_per_octet * i))) & 0xFF)
+        buf.append(octet)
+    return num_octets
+
+
+_CACHE_SIZE = 64
+
+
+class _FieldCache:
+    """Contains caches for small Int, UInt, VarInt, and VarUInt values.
+
+    For unsigned fields, values between 0 and 63 will have their representations cached. For signed fields, values
+    between -32 and 31 will have their representations cached. It is likely that a large proportion of subfields fit in
+    these ranges. For example:
+    - All values with lengths less than 64 bytes will retrieve their length subfield from the cache.
+    - All integer values with magnitudes less than 64 will retrieve their 'magnitude' subfields from the cache.
+    - Decimal values with exponents and/or coefficients within the signed cache range will retrieve those fields from
+      the cache.
+    - All timestamps will retrieve their month, day, hour, minute, and second fields from the cache (otherwise it would
+      be an invalid datetime). The fractional fields may be cached if they meet the decimal value requirement from the
+      previous bullet.
+    - The first 64 symbols will have their symbol ID fields retrieved from the cache. This applies to symbol values,
+      field names, and annotations.
+    - If a value has less than 64 bytes of annotations (hopefully this is true for all values), then its annotation
+      wrapper's 'annotation length' field will be cached.
+    """
```

Comment:
Yes. After looking at some profiling results generated by my (admittedly limited) test data, I found that field writing was taking the most time. The cache reduced the amount of time spent there by around 50%, leading to a modest 10% latency reduction. It's not huge, but I at least want to keep it around until we focus more on benchmarking, when we can revisit its merits if we wish. Using the rationale I put together in the comment, I convinced myself it would probably be widely beneficial, but I'm open to being proven wrong and reducing the size of the code base.